# Strings
<a href="https://colab.research.google.com/github/rambasnet/FDSPython-Notebooks/blob/master/Ch06-Strings.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

- http://openbookproject.net/thinkcs/python/english3e/strings.html
- string is a compound data type; sequence of characters
- can work with string as a single thing
- string variables are objects with their own attributes and methods
- help(str) to see all the methods
- commonly used methods: upper(), lower(), swapcase(), capitalize(), endswith(), isdigit(), find(), center(), count(), split(), etc.

In [ ]:
help(str)

In [1]:
ss = "Hello there beautiful World!"
tt = ss.upper()
print(tt)

HELLO THERE BEAUTIFUL WORLD!


In [2]:
print(ss)

Hello there beautiful World!


In [3]:
print(tt.capitalize())

Hello there beautiful world!


In [4]:
alist = tt.split()
print(alist)

['HELLO', 'THERE', 'BEAUTIFUL', 'WORLD!']


In [5]:
# examples of some methods
ss.count('o')

2

In [6]:
print(ss.swapcase())

hELLO THERE BEAUTIFUL wORLD!


## working with the parts of a string
- string can be sliced using [ index ] or [ inclStartIndex : ExclEndIndex : step] bracket operator
- negative indices are allowed
- len() built-in function gives the length of a string

In [14]:
# examples
s = "Pirates of the Caribbean"

In [16]:
print(s[1])

i


In [17]:
print(s[:])

Pirates of the Caribbean


In [18]:
print(s[1:])

irates of the Caribbean


In [19]:
# print just the pirates 
print(s[0:7])

Pirates


In [12]:
# print "the" from string s
theIndex = s.find("the")
print('the startst at', theIndex)
print(s[theIndex:theIndex+4])

the startst at 11
the 


In [13]:
# TODO
# print Caribbean from string s - hint use find function

In [ ]:
lastSpace = s.rfind(' ')
print(lastSpace)

In [21]:
# print the last character
print(s[-1])

n


In [22]:
# print string in reverse order
reversedS = s[-1::-1]

In [23]:
ss = "Pirates of the Caribbean."
print(ss[len(ss)-1])

.


In [24]:
# test whether a given string is palindrome
a= "racecar"
print ('palindrome') if a==a[::-1] else print('not palindrome')

palindrome


### string is immutable
- string objects/variables can't be modified in place!

In [1]:
a = 'hello'

In [2]:
a[0] = 'H'

TypeError: 'str' object does not support item assignment

## string traversal using for and while loop

In [26]:
# example for loop
# traversing string using index
for i in range(len(s)):
    print(s[i], end=' ')

P i r a t e s   o f   t h e   C a r i b b e a n 

In [25]:
# range-based loop traversing each character
for c in s:
    print(c, end=' ')

P i r a t e s   o f   t h e   C a r i b b e a n 

In [28]:
someStr = """afAdf@#456'"""

In [29]:
# example while loop
i = 0
while i < len(someStr):
    print(someStr[i], end=' ')
    i += 1

a f A d f @ # 4 5 6 ' 

## string comparison
- strings can be compared using comparison operators
- comparison operators are **==, !=, <=, >=, <, >**
- compares lexicographically (ASCII values)
- see [ASCII table](http://www.asciitable.com/)

In [2]:
# find ascii values of lower a and upper A
print(ord('A'), ord('a'))

65 97


In [33]:
# string comparison examples
print("apple" == "Apple")

False


In [34]:
print("apple" >= "ball")

False


In [35]:
print("apple" >= "Apple")

True


In [32]:
print("apple" == "Apple") # false

False


## in and not in operators
- help quickly test for membership

In [36]:
print("p" in "apple")

True


In [37]:
print("pe" in "apple")

False


In [38]:
print("aple" not in "apple")

True


## cleaning up strings
- e.g., remove punctuations
- traverse the string by removing any encountered punctation

In [39]:
# create a new string removing punctuations from the following string
ss = '"Well, I never did!", said Alice.'
print(ss)

"Well, I never did!", said Alice.


In [22]:
# one solution is to use ASCII value of each character
print(ord('a'))

97


In [40]:
newStr = ''
for c in ss:
    if ord(c) == ord(' '):
        newStr += c
    elif ord(c) >= ord('A') and ord(c) <= ord('Z'):
        newStr += c
    elif ord(c) >= ord('a') and ord(c) <= ord('z'):
        newStr += c
print(newStr)

Well I never did said Alice


In [41]:
# convert newStr to lowercase for case insensitive operations
newStr1 = newStr.lower()
print(newStr1)

well i never did said alice


In [42]:
# convert sentence into list of tokens/terms/words
words = newStr1.split()
print(words)

['well', 'i', 'never', 'did', 'said', 'alice']


In [43]:
# traverse through list of words
for w in words:
    print(w)

well
i
never
did
said
alice


In [44]:
# next solution using string library
import string
help(string)

Help on module string:

NAME
    string - A collection of string constants.

MODULE REFERENCE
    https://docs.python.org/3.7/library/string
    
    The following documentation is automatically generated from the Python
    source files.  It may be incomplete, incorrect or include features that
    are considered implementation detail and may vary between Python
    implementations.  When in doubt, consult the module reference at the
    location listed above.

DESCRIPTION
    Public module variables:
    
    whitespace -- a string containing all ASCII whitespace
    ascii_lowercase -- a string containing all ASCII lowercase letters
    ascii_uppercase -- a string containing all ASCII uppercase letters
    ascii_letters -- a string containing all ASCII letters
    digits -- a string containing all ASCII decimal digits
    hexdigits -- a string containing all ASCII hexadecimal digits
    octdigits -- a string containing all ASCII octal digits
    punctuation -- a string containing all

In [45]:
# string library has data that can be useful, e.g.
print(string.punctuation)

!"#$%&'()*+,-./:;<=>?@[\]^_`{|}~


In [46]:
ss

'"Well, I never did!", said Alice.'

In [48]:
newStr = ''
for c in ss:
    if c in string.ascii_lowercase:
        newStr += c
    elif c in string.ascii_uppercase:
        newStr += c
    elif c == ' ':
        newStr += ' '
        
print(newStr)

Well I never did said Alice


In [49]:
# write a function that removes all the punctations except for space
# returns new cleaned up string
def cleanUp(someStr):
    newStr = ''
    for c in someStr:
        if c.islower():
            newStr += c
        elif c.isupper():
            newStr += c
        elif c.isspace():
            newStr += c
    return newStr

In [50]:
s = cleanUp(ss)
print(s)

Well I never did said Alice


## string formatting
- format method
- use {} as replacement field
- numbers in curly braces are optional; determine which argument gets substituted
- each of the replace fields can also contain a format specification
    - < left alignment, > right, and ^ center, e.g. {1:<10}
    - 10 is width
    - type conversion such as f for float (.2f two decimal places), x for hex, etc.

In [44]:
name = "Arthur"
age = 25
s1 = "His name is {}!".format(name)
print(s1)

His name is Arthur!


In [46]:
# note age and name are provided in reverse order
print("His name is {1} and {1} is {0} years old.".format(age, name))

His name is Arthur and Arthur is 25 years old.


In [53]:
n1 = 4
n2 = 5.5
s3 = "{0} x {1} = {2} and {0} ^ {1} = {3:.2f}".format(n1, n2, n1*n2, n1**n2)
print(s3)

4 x 5.5 = 22.0 and 4 ^ 5.5 = 2048.00


In [49]:
n1 = "Paris"
n2 = "Whitney"
n3 = "Hilton"

print("Pi to three decimal places is {0:.3f}".format(3.1415926))
print("123456789 123456789 123456789 123456789 123456789 123456789")
print("|||{0:<15}|||{1:^15}|||{2:>15}|||Born in {3}|||"
        .format(n1,n2,n3,1981))
print("The decimal value {0} converts to hex value {0:x}"
        .format(123456))

Pi to three decimal places is 3.142
123456789 123456789 123456789 123456789 123456789 123456789
|||Paris          |||    Whitney    |||         Hilton|||Born in 1981|||
The decimal value 123456 converts to hex value 1e240


In [50]:
letter = """
Dear {0} {2}.
 {0}, I have an interesting money-making proposition for you!
 If you deposit $10 million into my bank account, I can
 double your money ...
"""

print(letter.format("Paris", "Whitney", "Hilton"))
print(letter.format("Bill", "Warren", "Jeff"))


Dear Paris Hilton.
 Paris, I have an interesting money-making proposition for you!
 If you deposit $10 million into my bank account, I can
 double your money ...


Dear Bill Jeff.
 Bill, I have an interesting money-making proposition for you!
 If you deposit $10 million into my bank account, I can
 double your money ...



In [51]:
layout = "{0:>4}{1:>6}{2:>6}{3:>8}{4:>13}{5:>24}"

print(layout.format("i", "i**2", "i**3", "i**5", "i**10", "i**20"))
for i in range(1, 11):
    print(layout.format(i, i**2, i**3, i**5, i**10, i**20))

   i  i**2  i**3    i**5        i**10                   i**20
   1     1     1       1            1                       1
   2     4     8      32         1024                 1048576
   3     9    27     243        59049              3486784401
   4    16    64    1024      1048576           1099511627776
   5    25   125    3125      9765625          95367431640625
   6    36   216    7776     60466176        3656158440062976
   7    49   343   16807    282475249       79792266297612001
   8    64   512   32768   1073741824     1152921504606846976
   9    81   729   59049   3486784401    12157665459056928801
  10   100  1000  100000  10000000000   100000000000000000000


## Exercises

Exercise 1. print a neat looking multiplication table like this:
<pre>
        1   2   3   4   5   6   7   8   9  10  11  12
  :--------------------------------------------------
 1:     1   2   3   4   5   6   7   8   9  10  11  12
 2:     2   4   6   8  10  12  14  16  18  20  22  24
 3:     3   6   9  12  15  18  21  24  27  30  33  36
 4:     4   8  12  16  20  24  28  32  36  40  44  48
 5:     5  10  15  20  25  30  35  40  45  50  55  60
 6:     6  12  18  24  30  36  42  48  54  60  66  72
 7:     7  14  21  28  35  42  49  56  63  70  77  84
 8:     8  16  24  32  40  48  56  64  72  80  88  96
 9:     9  18  27  36  45  54  63  72  81  90  99 108
10:    10  20  30  40  50  60  70  80  90 100 110 120
11:    11  22  33  44  55  66  77  88  99 110 121 132
12:    12  24  36  48  60  72  84  96 108 120 132 144
</pre>

Exercise 2. Write a program that determines whether a given string is palindrome. Palindrome is a word, phrase, or sequence that reads the same backward as forward, e.g., madam or nurses run or race car.

Exercise 2.1. Convert Exercise 2 into a function and write at least two test cases.

Exercise 3. Write a program that calculates number of trials required to guess a 3 digit pass code 777 (starting from 000, 001, 002, 003, 004, 005..., 010, etc.) using some brute force technique.

Exercise 3.1. Convert Exercise 3 into a function and write at least 3 test cases.

## Kattis problems involving string
1. Avion - https://open.kattis.com/problems/avion
- Apaxiaans - https://open.kattis.com/problems/apaxiaaans
- Hissing Microphone - https://open.kattis.com/problems/hissingmicrophone
- Reversed Binary Numbers - https://open.kattis.com/problems/reversebinary
- Kemija - https://open.kattis.com/problems/kemija08
- Simon Says - https://open.kattis.com/problems/simonsays
- Simon Says - https://open.kattis.com/problems/simon
- Quite a Problem - https://open.kattis.com/problems/quiteaproblem
- Eligibility - https://open.kattis.com/problems/eligibility
- Charting Progress - https://open.kattis.com/problems/chartingprogress
- Pig Latin - https://open.kattis.com/problems/piglatin
- Battle Simulation - https://open.kattis.com/problems/battlesimulation
- Palindromic Password - https://open.kattis.com/problems/palindromicpassword
- Image Decoding - https://open.kattis.com/problems/imagedecoding